File: mac/daemon/src/routes/connection-diagnostics.ts
// `GET /api/connection-diagnostics` — 서브시스템 «읽기 전용» 연결 스냅샷.
//
// (별개 기능과의 구분: `/api/diagnostics`(routes/diagnostics.ts) 는 «문제 신고/진단 번들»
//  — crash 마커 + 마스킹된 로그 tail. 여기 «연결 진단» 은 Tor/sshd/디스크/에이전트 CLI 의
//  서브시스템 «상태» 를 안정적 코드로 분류해 iOS 「연결 진단」 화면이 원인·조치를 안내하게 한다.)
//
// 동기: 연결이 실패해도 daemon 의 /health 는 {ok,time,connectedClients} 뿐이라 Tor/sshd/디스크/
// 에이전트 CLI 상태를 노출하지 못했고, 에러는 구조화 코드 없이 문자열뿐이라 iOS 가 「왜·무엇을
// 하라」 를 안내할 수 없었다. 이 엔드포인트는 각 서브시스템의 상태를 안정적 코드(connection-
// diagnostics/codes.ts) + 원시 지표로 내보내 iOS 진단 화면이 사람이 읽는 localize 문구·권장 조치로 매핑한다.
//
// 경계: daemon 은 UI 표면이 없으므로 «사용자 문구» 를 만들지 않는다 — 코드(식별자)·숫자·boolean
// 만 내보낸다. 색·문구·시각 포맷은 전부 iOS 가 한다. 외부 IPv4 «값» 은 싣지 않고 존재 여부
// (boolean)만 — 로그의 `secret.*` redact 정책과 같은 취지(주소 자체는 진단에 불필요).
//
// 모든 I/O(프로세스 생존·디스크 statfs·DB count 등)는 deps 로 «주입» 받아, 분류 로직과 응답
// 계약을 실제 시스템 없이 테스트할 수 있게 한다 (connection-diagnostics.test.ts).

import { Hono } from "hono";
import { bearerAuth } from "../auth.js";
import {
  classifyTor,
  classifySsh,
  classifyReachability,
  classifyAgentCli,
  classifyDisk,
  classifyLogs,
  worstLevel,
  type DiagnosticLevel,
  type DiagnosticCode,
} from "../connection-diagnostics/codes.js";

/** 서브시스템 식별자 — iOS 가 표시명·아이콘으로 매핑한다(번역 대상은 표시명, id 는 식별자). */
export type DiagnosticSubsystemId =
  | "tor"
  | "sshd"
  | "reachability"
  | "agent_cli"
  | "disk"
  | "logs"
  | "network";

/**
 * 표시용 원시 지표 — 전부 옵셔널(서브시스템마다 관련 필드만 채운다). iOS 가 사람이 읽는 형식
 * (바이트→사람 단위, epoch→날짜)으로 포맷한다. 여기엔 사용자 문구를 담지 않는다.
 */
export type DiagnosticMetrics = {
  torProcessAlive?: boolean;
  torBootstrapPercent?: number;
  onionPublished?: boolean;
  sshListening?: boolean;
  sshPort?: number;
  diskFreeBytes?: number;
  diskTotalBytes?: number;
  unifiedLogBytes?: number;
  ptyChunkCount?: number;
  lanOnly?: boolean;
  lanCandidateCount?: number;
  externalIPv4Present?: boolean;
  ipFetchedAt?: number;
  lastIpChangeAt?: number;
  lastReconnectAt?: number;
};

/** 서브시스템 안의 개별 항목(예: 에이전트 CLI 하나). */
export type DiagnosticItem = {
  id: string;
  /** 사람 친화 이름(예: "Claude Code") — 식별자성. iOS 가 verbatim 표시. */
  label: string;
  level: DiagnosticLevel;
  code: DiagnosticCode;
  /** 코드성 보조 문자열(예: 설치 명령 installHint) — 번역 대상 아님. */
  detail?: string;
};

export type DiagnosticSubsystem = {
  id: DiagnosticSubsystemId;
  level: DiagnosticLevel;
  code: DiagnosticCode;
  metrics?: DiagnosticMetrics;
  items?: DiagnosticItem[];
};

export type DiagnosticsResponse = {
  v: 1;
  /** 스냅샷 생성 시각 (epoch ms). */
  generatedAt: number;
  /** 서브시스템 중 가장 나쁜 심각도. */
  overall: DiagnosticLevel;
  subsystems: DiagnosticSubsystem[];
};

/** 스냅샷 수집에 필요한 모든 관측 신호 — server.ts 가 실제 핸들/모듈로 채운다. */
export type DiagnosticsDeps = {
  getTorProcessAlive: () => boolean;
  /** 0~100. 100 이면 부트스트랩 완료. */
  getTorBootstrapPercent: () => number;
  /** onion 주소 (게시됨 신호). null 이면 미게시. */
  getOnionAddress: () => string | null;
  getSshListening: () => boolean;
  getSshPort: () => number;
  getDiskFree: () => { freeBytes: number; totalBytes: number } | null;
  listAgentCli: () => Array<{
    id: string;
    displayName: string;
    detected: boolean;
    installHint?: string;
  }>;
  isLanOnly: () => boolean;
  getLanCandidateCount: () => number;
  getUnifiedLogBytes: () => number | null;
  getPtyChunkCount: () => number | null;
  /** 외부 IPv4 (있음/없음만 노출 — 값 자체는 싣지 않는다). */
  getExternalIPv4: () => string | null;
  getIpFetchedAt: () => number | null;
  getLastIpChangeAt: () => number | null;
  getLastReconnectAt: () => number | null;
  now: () => number;
};

/**
 * 관측 신호 → 진단 스냅샷. 순수(주입 deps 의 결과만 사용) — 분류·집계를 실제 I/O 없이 테스트.
 */
export function buildDiagnosticsSnapshot(deps: DiagnosticsDeps): DiagnosticsResponse {
  const subsystems: DiagnosticSubsystem[] = [];

  // Tor — 부트스트랩%·onion 게시 여부.
  const torProcessAlive = deps.getTorProcessAlive();
  const torPct = deps.getTorBootstrapPercent();
  const onion = deps.getOnionAddress();
  const torVerdict = classifyTor({
    processAlive: torProcessAlive,
    bootstrapped: torPct >= 100,
    onionPresent: onion != null,
  });
  subsystems.push({
    id: "tor",
    level: torVerdict.level,
    code: torVerdict.code,
    metrics: {
      torProcessAlive,
      torBootstrapPercent: torPct,
      onionPublished: onion != null,
    },
  });

  // sshd — listening 여부 + 포트.
  const sshListening = deps.getSshListening();
  const sshVerdict = classifySsh({ processAlive: sshListening });
  subsystems.push({
    id: "sshd",
    level: sshVerdict.level,
    code: sshVerdict.code,
    metrics: { sshListening, sshPort: deps.getSshPort() },
  });

  // 외부 연결성 — LAN 전용 정책 + LAN 후보 수.
  const lanOnly = deps.isLanOnly();
  const lanCount = deps.getLanCandidateCount();
  const reachVerdict = classifyReachability({ lanOnly, lanCandidateCount: lanCount });
  subsystems.push({
    id: "reachability",
    level: reachVerdict.level,
    code: reachVerdict.code,
    metrics: { lanOnly, lanCandidateCount: lanCount },
  });

  // 에이전트 CLI — 등록된 어댑터별 탐지 결과. 서브시스템 level/code = 항목 중 가장 나쁜 것.
  const agents = deps.listAgentCli();
  const items: DiagnosticItem[] = agents.map((a) => {
    const v = classifyAgentCli({ detected: a.detected });
    return {
      id: a.id,
      label: a.displayName,
      level: v.level,
      code: v.code,
      // 미설치일 때만 설치 명령(installHint)을 보조로 — 코드성 문자열, iOS 가 monospace 로 표시.
      ...(a.detected ? {} : a.installHint ? { detail: a.installHint } : {}),
    };
  });
  subsystems.push({
    id: "agent_cli",
    level: worstLevel(items.map((i) => i.level)),
    code: items.find((i) => i.code !== "ok")?.code ?? "ok",
    items,
  });

  // 디스크 여유.
  const disk = deps.getDiskFree();
  const diskVerdict = classifyDisk({ freeBytes: disk?.freeBytes ?? null });
  subsystems.push({
    id: "disk",
    level: diskVerdict.level,
    code: diskVerdict.code,
    metrics: disk ? { diskFreeBytes: disk.freeBytes, diskTotalBytes: disk.totalBytes } : {},
  });

  // 로그 — unified.log / pty_chunk 크기.
  const unifiedLogBytes = deps.getUnifiedLogBytes();
  const ptyChunkCount = deps.getPtyChunkCount();
  const logsVerdict = classifyLogs({ unifiedLogBytes, ptyChunkCount });
  subsystems.push({
    id: "logs",
    level: logsVerdict.level,
    code: logsVerdict.code,
    metrics: {
      ...(unifiedLogBytes != null ? { unifiedLogBytes } : {}),
      ...(ptyChunkCount != null ? { ptyChunkCount } : {}),
    },
  });

  // 네트워크 — 정보성(외부 IP 존재·마지막 IP변경/재연결 시각). 실패 코드 없음(ok).
  const ipv4 = deps.getExternalIPv4();
  const ipFetchedAt = deps.getIpFetchedAt();
  const lastIpChangeAt = deps.getLastIpChangeAt();
  const lastReconnectAt = deps.getLastReconnectAt();
  subsystems.push({
    id: "network",
    level: "ok",
    code: "ok",
    metrics: {
      externalIPv4Present: ipv4 != null,
      ...(ipFetchedAt != null ? { ipFetchedAt } : {}),
      ...(lastIpChangeAt != null ? { lastIpChangeAt } : {}),
      ...(lastReconnectAt != null ? { lastReconnectAt } : {}),
    },
  });

  return {
    v: 1,
    generatedAt: deps.now(),
    overall: worstLevel(subsystems.map((s) => s.level)),
    subsystems,
  };
}

/** `/api/connection-diagnostics` 라우트 — 다른 `/api/*` 와 동일하게 bearer 인증. */
export function connectionDiagnosticsRoute(deps: DiagnosticsDeps) {
  const app = new Hono();
  app.use("*", bearerAuth);
  app.get("/", (c) => c.json(buildDiagnosticsSnapshot(deps)));
  return app;
}

File: ios/PocketSisyphus/Models/ConnectionDiagnostics.swift
import SwiftUI

/// daemon `GET /api/diagnostics` 의 서브시스템 «읽기 전용» 진단 스냅샷 — `mac/daemon/src/routes/
/// diagnostics.ts` 의 `DiagnosticsResponse` 와 1:1.
///
/// 경계: daemon 은 UI 표면이 없어 «사용자 문구» 를 만들지 않는다 — 안정적 코드(식별자)·원시 숫자·
/// boolean 만 내보낸다. 색·사람이 읽는 제목·권장 조치·시각/바이트 포맷은 «전부» 여기(iOS)서 한다.
/// 색은 이 레포의 «의미» 약속을 따른다(DesignTokens 의 색상 정책):
///   ok=success(초록) · warning=warning(노랑) · error=danger(빨강) · unknown=secondary(중립·자동 적응).
/// warning(노랑)↔pro(주황) 혼동 금지 — 진단은 프리미엄이 아니라 pro(주황)를 쓰지 않는다.
struct DiagnosticsResponse: Codable, Equatable {
    let v: Int
    /// 스냅샷 생성 시각 (epoch ms).
    let generatedAt: Int64
    /// 서브시스템 중 가장 나쁜 심각도 ("ok" | "warning" | "error" | "unknown").
    let overall: String
    let subsystems: [DiagnosticSubsystem]

    var overallLevel: DiagnosticLevel { DiagnosticLevel(apiValue: overall) }
}

/// 한 서브시스템(Tor·sshd·디스크 등)의 상태. `id` 는 식별자(번역 대상 아님) — iOS 가 표시명으로 매핑.
struct DiagnosticSubsystem: Codable, Equatable, Identifiable {
    /// "tor" | "sshd" | "reachability" | "agent_cli" | "disk" | "logs" | "network". 식별자.
    let id: String
    let level: String
    let code: String
    let metrics: DiagnosticMetrics?
    /// 서브시스템 안의 개별 항목 (예: 에이전트 CLI 하나).
    let items: [DiagnosticItem]?

    var levelEnum: DiagnosticLevel { DiagnosticLevel(apiValue: level) }
    var codeEnum: DiagnosticCode { DiagnosticCode(apiValue: code) }

    /// 서브시스템 표시명 — id(식별자)를 사람이 읽는 localize 문구로. 미지 id 는 그대로 노출.
    var displayName: String {
        switch id {
        case "tor": return String(localized: "Tor 네트워크")
        case "sshd": return String(localized: "SSH 서버")
        case "reachability": return String(localized: "외부 연결성")
        case "agent_cli": return String(localized: "에이전트 CLI")
        case "disk": return String(localized: "디스크 공간")
        case "logs": return String(localized: "로그")
        case "network": return String(localized: "네트워크 · IP")
        default: return id
        }
    }

    /// SF Symbol — 서브시스템 식별 아이콘 (상태 아이콘과 별개의 «무엇» 표시).
    var symbol: String {
        switch id {
        case "tor": return "globe.americas"
        case "sshd": return "terminal"
        case "reachability": return "wifi"
        case "agent_cli": return "cpu"
        case "disk": return "internaldrive"
        case "logs": return "doc.text"
        case "network": return "network"
        default: return "questionmark"
        }
    }
}

/// 서브시스템 안의 개별 항목 — 에이전트 CLI 하나처럼 여러 개가 묶이는 경우.
struct DiagnosticItem: Codable, Equatable, Identifiable {
    let id: String
    /// 사람 친화 이름(예: "Claude Code") — 식별자성. verbatim 표시.
    let label: String
    let level: String
    let code: String
    /// 코드성 보조 문자열(예: 설치 명령 installHint) — 번역 대상 아님, monospace 표시.
    let detail: String?

    var levelEnum: DiagnosticLevel { DiagnosticLevel(apiValue: level) }
    var codeEnum: DiagnosticCode { DiagnosticCode(apiValue: code) }
}

/// 표시용 원시 지표 — 전부 옵셔널(서브시스템마다 관련 필드만 채워짐). iOS 가 사람이 읽는
/// 형식(바이트→사람 단위, epoch→날짜)으로 포맷한다. 모르는 키는 nil(additive-safe).
struct DiagnosticMetrics: Codable, Equatable {
    var torProcessAlive: Bool? = nil
    var torBootstrapPercent: Int? = nil
    var onionPublished: Bool? = nil
    var sshListening: Bool? = nil
    var sshPort: Int? = nil
    var diskFreeBytes: Int64? = nil
    var diskTotalBytes: Int64? = nil
    var unifiedLogBytes: Int64? = nil
    var ptyChunkCount: Int? = nil
    var lanOnly: Bool? = nil
    var lanCandidateCount: Int? = nil
    var externalIPv4Present: Bool? = nil
    var ipFetchedAt: Int64? = nil
    var lastIpChangeAt: Int64? = nil
    var lastReconnectAt: Int64? = nil
}

// MARK: - 심각도 → 색·아이콘·라벨 (이 레포 색상 약속의 SSOT 매핑)

/// 진단 심각도. 색은 «의미» 약속 그대로 — 장식·강조로 status 색을 빌려 쓰지 않는다.
enum DiagnosticLevel: String {
    case ok
    case warning
    case error
    case unknown

    /// 모르는 값은 unknown 으로 — 미래의 daemon 이 새 level 을 보내도 안전하게 폴백.
    init(apiValue: String) { self = DiagnosticLevel(rawValue: apiValue) ?? .unknown }

    /// 상태색 — 본문/배경은 .primary/.secondary 를 쓰고, 이 tint 는 «상태 아이콘/배지» 전용.
    var tint: Color {
        switch self {
        case .ok: return Theme.success        // 초록
        case .warning: return Theme.warning    // 노랑 (주의/설정 필요)
        case .error: return Theme.danger       // 빨강 (실패)
        case .unknown: return .secondary       // 중립·자동 적응 (정보·보조)
        }
    }

    var icon: String {
        switch self {
        case .ok: return "checkmark.circle.fill"
        case .warning: return "exclamationmark.triangle.fill"
        case .error: return "xmark.octagon.fill"
        case .unknown: return "questionmark.circle"
        }
    }

    /// 사람이 읽는 상태 라벨 — 상태 아이콘의 accessibilityLabel + 배지 텍스트로 쓴다.
    var localizedLabel: String {
        switch self {
        case .ok: return String(localized: "정상")
        case .warning: return String(localized: "주의 필요")
        case .error: return String(localized: "오류")
        case .unknown: return String(localized: "확인 불가")
        }
    }
}

// MARK: - 안정적 코드 → 사람이 읽는 제목·권장 조치 (localize)

/// 연결/런타임 실패의 안정적 식별자 — `mac/daemon/src/diagnostics/codes.ts` 의 `DiagnosticCode`
/// 와 손으로 짝지어 관리한다. daemon 은 이 코드만 내보내고, 「왜·무엇을 하라」 문구는 여기서 매핑.
enum DiagnosticCode: String {
    case ok
    case unknown
    case torProcessDown = "tor_process_down"
    case torNotBootstrapped = "tor_not_bootstrapped"
    case torDescriptorMissing = "tor_descriptor_missing"
    case sshNotListening = "ssh_not_listening"
    case sshHostkeyMismatch = "ssh_hostkey_mismatch"
    case lanBlockedNoPublicFallback = "lan_blocked_no_public_fallback"
    case agentCliMissing = "agent_cli_missing"
    case diskLow = "disk_low"
    case diskCritical = "disk_critical"
    case logOversized = "log_oversized"

    /// 모르는 코드는 unknown 으로 폴백 — 미래 daemon 의 새 코드도 안전하게 처리.
    init(apiValue: String) { self = DiagnosticCode(rawValue: apiValue) ?? .unknown }

    /// 사람이 읽는 «왜» 한 줄. ok 는 nil(정상은 상태 라벨로 충분).
    var localizedTitle: String? {
        switch self {
        case .ok: return nil
        case .unknown: return String(localized: "상태를 확인할 수 없어요")
        case .torProcessDown: return String(localized: "Tor가 실행되고 있지 않아요")
        case .torNotBootstrapped: return String(localized: "Tor 연결을 준비하는 중이에요")
        case .torDescriptorMissing: return String(localized: "Tor 주소를 게시하지 못했어요")
        case .sshNotListening: return String(localized: "SSH 서버가 응답하지 않아요")
        case .sshHostkeyMismatch: return String(localized: "서버 호스트 키가 바뀌었어요")
        case .lanBlockedNoPublicFallback: return String(localized: "연결할 사설 주소가 없어요 (LAN 전용)")
        case .agentCliMissing: return String(localized: "에이전트 CLI가 설치돼 있지 않아요")
        case .diskLow: return String(localized: "디스크 여유 공간이 부족해요")
        case .diskCritical: return String(localized: "디스크 공간이 매우 부족해요")
        case .logOversized: return String(localized: "로그가 비정상적으로 커졌어요")
        }
    }

    /// 권장 조치 — 「무엇을 하라」. 없으면 nil.
    var localizedAction: String? {
        switch self {
        case .ok, .unknown: return nil
        case .torProcessDown:
            return String(localized: "Mac 앱을 재시작하세요. 자동 복구를 시도하지만 계속되면 재시작이 필요해요.")
        case .torNotBootstrapped:
            return String(localized: "Tor 회로를 만드는 중이에요. 잠시 기다리면 자동으로 연결돼요.")
        case .torDescriptorMissing:
            return String(localized: "잠시 기다린 뒤 다시 시도하세요. 문제가 계속되면 Mac 앱을 재시작하거나 페어링을 다시 하세요.")
        case .sshNotListening:
            return String(localized: "Mac 앱을 재시작하세요. 포트가 다른 프로그램에 점유됐다면 Mac의 포트 설정에서 바꿀 수 있어요.")
        case .sshHostkeyMismatch:
            return String(localized: "Mac을 재설치했다면 페어링을 다시 하세요. 그렇지 않다면 서버 위장(중간자)일 수 있어 직접 연결을 차단했어요.")
        case .lanBlockedNoPublicFallback:
            return String(localized: "같은 Wi‑Fi에 연결하거나, 설정에서 LAN 전용 모드를 끄면 외부 경로로 연결할 수 있어요.")
        case .agentCliMissing:
            return String(localized: "아래 설치 명령으로 에이전트 CLI를 설치하면 그 에이전트로 세션을 만들 수 있어요.")
        case .diskLow:
            return String(localized: "저장 공간을 확보하세요. 로그·세션 기록 저장이 실패할 수 있어요.")
        case .diskCritical:
            return String(localized: "지금 저장 공간을 확보하세요. 세션 기록과 로그가 저장되지 않을 수 있어요.")
        case .logOversized:
            return String(localized: "동작에는 문제가 없지만, 로그 정리를 권장해요.")
        }
    }
}
